### PR TITLE
server: do not deploy or upgrade vm with inactive service offering

### LINKED ISSUE #6740: API: VM Created from deleted compute offering
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
API
~~~

##### CLOUDSTACK VERSION

~~~
4.16.1
~~~

##### CONFIGURATION

Compute offerings
##### OS / ENVIRONMENT

CentOS7

##### SUMMARY

VM can be created from a deleted offering via API

##### STEPS TO REPRODUCE

~~~
Create a compute offering and note the UUID
Remove the Compute offering
Make sure its removed from UI
Create a instance using the UUID of the deleted Offering
Instance Get created VIA compute offering
~~~

##### EXPECTED RESULTS

~~~
Instance should not get created via API using deleted offering
~~~

##### ACTUAL RESULTS

~~~
Instance got created VIA deleted compute offering
~~~

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes #6740 

The deleted service offering is not removed , but changed to Inactive state.
Therefore the state needs to be checked when deploy/create new vm or upgrade vm.
Similar check is added to router service offering as well.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

```
    (localcloud) 🐱 > deploy virtualmachine zoneid=7c9769a1-ba70-4f3a-a976-5a6ae7d946bb templateid=936f5eda-6be2-11ed-a030-1e0066000350 networkids=e7d3308c-03fc-4411-998d-ff45a2c1fccc  serviceofferingid=c1ebd091-57a6-427a-99a9-6c482de84a4f
    {
      "virtualmachine": {
        "account": "admin",
    ...
      }
    }

    (localcloud) 🐱 > delete serviceoffering id=c1ebd091-57a6-427a-99a9-6c482de84a4f
    {
      "success": true
    }

    (localcloud) 🐱 > deploy virtualmachine zoneid=7c9769a1-ba70-4f3a-a976-5a6ae7d946bb templateid=936f5eda-6be2-11ed-a030-1e0066000350 networkids=e7d3308c-03fc-4411-998d-ff45a2c1fccc  serviceofferingid=c1ebd091-57a6-427a-99a9-6c482de84a4f
    🙈 Error: (HTTP 431, error code 4350) service offering is Inactive: c1ebd091-57a6-427a-99a9-6c482de84a4f
```

### CODE REVIEW

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 1]
```diff
@@ -3741,6 +3741,10 @@ public void checkIfCanUpgrade(final VirtualMachine vmInstance, final ServiceOffe
             throw new InvalidParameterValueException("Invalid parameter, newServiceOffering can't be null");
         }
 
+        if (ServiceOffering.State.Inactive.equals(newServiceOffering.getState())) {
+            throw new InvalidParameterValueException("New service offering is Inactive: " + newServiceOffering.getUuid());
```
Comment (stephankruggg) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:3745]:
```suggestion
            throw new InvalidParameterValueException(String.format("New service offering is inactive: [%s].", newServiceOffering.getUuid()));
```

**Review by stephankruggg — COMMENTED** [review]
Hi @weizhouapache, can you extract the inactive check to a method and reuse it in both scenarios to avoid code duplication?

**Review by rohityadavcloud — APPROVED** [review]

**Comment by rohityadavcloud** [comment]
Is this bugfix good for 4.18 cc @weizhouapache @DaanHoogland ?
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
> Is this bugfix good for 4.18 cc @weizhouapache @DaanHoogland ? @blueorangutan package

@rohityadavcloud cc @DaanHoogland 
yes but it has low priority

**Comment by weizhouapache** [comment]
@blueorangutan package

**Review by vladimirpetrov — APPROVED** [review]
LGTM based on manual testing.

**Pull request merged**